### PR TITLE
[SPARK-22335][SQL] Clarify union behavior on Dataset of typed objects in the document

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Seems that end users can be confused by the union's behavior on Dataset of typed objects. We can clarity it more in the document of `union` function.

## How was this patch tested?

Only document change.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]
Yea, I have been there in many JIRAs related with it too. Describing the current behaviour. LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -1753,6 +1753,27 @@ class Dataset[T] private[sql](
    *
    * Also as standard in SQL, this function resolves columns by position (not by name).
    *
+   * Notice that the column positions in the schema aren't necessarily matched with the
+   * fields in the typed objects in a Dataset. This function resolves columns by their positions
+   * in the schema, not the fields in the typed objects:
+   *
+   * {{{
+   *   case class Test(a : String, b : String)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: `(a : String, b : String)` -> `(a: String, b: String)` maybe?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -1753,6 +1753,27 @@ class Dataset[T] private[sql](
    *
    * Also as standard in SQL, this function resolves columns by position (not by name).
    *
+   * Notice that the column positions in the schema aren't necessarily matched with the
+   * fields in the typed objects in a Dataset. This function resolves columns by their positions
+   * in the schema, not the fields in the typed objects:
+   *
+   * {{{
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Could we clarify this is a Scala example?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Yes, thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Sure.

**Comment by viirya** [comment]
@HyukjinKwon Thanks for reviewing.

**Comment by HyukjinKwon** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -1753,6 +1753,27 @@ class Dataset[T] private[sql](
...
+   *   val ds2 = Seq(("b", "a")).toDF("b", "a").as[Test] // ds2's schema: [b: String, a: String]
+   *   ds1.union(ds2).show
+   *
+   *   // output:
+   *   // +---+---+
+   *   // |  a|  b|
+   *   // +---+---+
+   *   // |  a|  b|
+   *   // |  b|  a|
+   *   // +---+---+
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Please use the same example as `union `. Just need to add a comment to explain it is also applicable to the strongly-typed JVM objects.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Sorry, I don't get the meaning of same example as `union`. This is the only example of `union`, if I don't miss anything.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Added a comment to say it's also applicable to the strongly-typed JVM objects.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Please use the same example as `union`. Users can easily get the difference between `union` and `unionByName`.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Sorry, but this is the example of `union`. Which example you mean to use here? Are you meaning to use the example of `unionByName`?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: If I understand your comment correctly, you mean we don't need to add this example for explaining how `union` works on typed objects. Just a normal union example is good.

I'm ok for it, although from past experiences, there are many end users getting confused by this difference.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by viirya** [comment]
Thanks @HyukjinKwon @gatorsmile

**Pull request closed without merging**